AndroidSDKFcm/src/main/java/com/leanplum/LeanplumPushFirebaseMessagingService.java

```diff
@@ -86,4 +99,63 @@ private Bundle getBundle(Map<String, String> messageMap) {
     }
     return bundle;
   }
+
+  private Boolean isDuplicateNotification(Map<String, String> messageMap) {
+    Boolean isDuplicate = false;
+    String messageId = "";
+    if (messageMap.containsKey("_lpn")) {
+      messageId = messageMap.get("_lpn");
+    }
+    if (messageMap.containsKey("_lpm")) {
+      messageId = messageMap.get("_lpm");
+    }
+
+    Map<String, Object> handledNotifications = retrieveHandledNotifications();
+    handledNotifications = purgeExpiredHandledNotifications(handledNotifications);
+    if (handledNotifications.containsKey(messageId)) {
+      isDuplicate = true;
+    } else {
+      handledNotifications.put(messageId, new Date().getTime());
+    }
+    persistHandledNotifications(handledNotifications);
+    return isDuplicate;
+  }
+
+  private Map<String, Object> retrieveHandledNotifications() {
+    SharedPreferences preferences = this.getSharedPreferences(Constants.Defaults.LEANPLUM,
+            Context.MODE_PRIVATE);
+    String jsonString = preferences.getString(handledNotificationsKey(), null);
+    Map<String, Object> handledNotifications;
+    if (jsonString != null) {
+      handledNotifications = JsonConverter.fromJson(jsonString);
+    } else {
+      handledNotifications = new HashMap<>();
+    }
+    return handledNotifications;
+  }
+
+  private void persistHandledNotifications(Map<String, Object> handledNotifications) {
```

Comment:
Definitely VarCache needs to be used. This code should not be in this file of firebase.